### PR TITLE
mobile app background improvements CORE-5106

### PR DESCRIPTION
**Pull request by mmaxim**
This patch aims to fix the weirdness with processing iOS background notifications. The major change is to plumb the state of the mobile app into the service so we can respond to state changes smarter, and to behave different when in the background (for chat at least). The details are the following:

1.) Create a new type, `AppState`, which tracks the state of the app. A key method of this type is `NextUpdate`, which returns a channel which will yield the next state of the app. Clients can wait on this channel and react accordingly (like `chat.Syncer`).
2.) Create a a new RPC protocol for `AppState` and hook it up in service.
3.) Change `Syncer` to not send stale notifications when the app is not in the foreground. The UI behaves unpredictably in this model, and it is safer to simply not get it involved while executing in the background. By just ignoring `sendNotificationOnce` in `Syncer`, we can just take advantage of the queuing system already there.
4.) Change `PushHandler` to not send any `NotifyRouter` calls when not in the foreground. Instead, upon reception of a Gregor message, it will send a thread stale notifications to `Syncer`. This will get queued in the background, and blasted out when the app is foregrounded.
5.) Change data protection mode to `NSFileProtectionCompleteUntilFirstUserAuthentication`, anything stronger simply doesn't work with LevelDB.
6.) Disconnect from Gregor on backgrounding, and reconnect on foreground. This is so that we don't get backgrounded in the middle of a ping, or a timer wait and add a ton of non-determinism to the foregrounding.
7.) Plumb the `AppState` from React into the service. @chrisnojima I think I did this right, but let me know if something is off.

cc @maxtaco (just so you are in the loop, no need to review if you are busy).

### CODE REVIEW

File: go/service/gregor.go [hunk 1]
```diff
@@ -194,24 +193,52 @@ func newGregorHandler(g *globals.Context) *gregorHandler {
...
+		case keybase1.AppState_FOREGROUND:
+			g.Shutdown()
+			g.chatLog.Debug(context.Background(), "foregrounded, reconnecting")
+			if err := g.Connect(g.uri); err != nil {
+				g.chatLog.Debug(context.Background(), "error reconnecting")
+			}
+		case keybase1.AppState_INACTIVE:
+			g.chatLog.Debug(context.Background(), "backgrounded, shutting down connection")
+			g.Shutdown()
+		}
```
Comment (patrickxb) [inline go/service/gregor.go]: Is there anything needed for `AppState_BACKGROUND`?

Comment (mmaxim) [inline go/service/gregor.go]: No, you always get `INACTIVE` before that, which for managing the connection is all you need.

Comment (patrickxb) [inline go/service/gregor.go]: Ok.

File: shared/actions/app.js [hunk 2]
```diff
@@ -13,8 +18,41 @@ function hideKeyboard (): Constants.HideKeyboard {
   return {payload: undefined, type: 'app:hideKeyboard'}
 }
 
+function mobileAppStateChanged (nextAppState: string): Constants.MobileAppState {
+  return {payload: {nextAppState}, type: 'app:mobileAppState'}
+}
+
+function * _onMobileAppStateChanged (action : Constants.MobileAppState): SagaGenerator<any, any> {
+  let appState = Types.AppStateAppState.foreground
+  const nextAppState = action.payload.nextAppState
```
Comment (chrisnojima) [inline shared/actions/app.js]: I'd likely write this as

```
const nextAppState = action.payload.nextAppState

const focusState = {
    active: true,
    inactive: false,
    background: false,
}[nextAppState]

yield put(changedFocus(focusState))

const state = {
    active: Types.AppStateAppState.foreground,
    inactive: Types.AppStateAppState.inactive,
    background: Types.AppStateAppState.background,
}[nextAppState] || Types.AppStateAppState.foreground

yield call(Types.appStateUpdateAppStateRpc, {param: {state}})
```

but you can leave it

File: shared/actions/push/index.js [hunk 3]
```diff
@@ -40,6 +42,15 @@ function * pushNotificationSaga (notification: Constants.PushNotification): Saga
     // wake up and do what it needs to do.
     if (payload.data && payload.data.silent) {
       console.info('Push notification: silent notification received')
+      const appFocused = yield select(Shared.focusedSelector)
+      if (!appFocused) {
+        console.info('Push notification: sending state update to service')
+        yield call(Types.appStateUpdateAppStateRpc, {
+          param: {
+            state: Types.AppStateAppState.backgroundactive,
```
Comment (chrisnojima) [inline shared/actions/push/index.js]: we should chat about this in person

File: shared/constants/app.js [hunk 4]
```diff
@@ -4,5 +4,6 @@ import type {NoErrorTypedAction} from './types/flux'
 export type ChangedFocus = NoErrorTypedAction<'app:changedFocus', {appFocused: boolean}>
 export type AppLink = NoErrorTypedAction<'app:link', {link: string}>
 export type HideKeyboard = NoErrorTypedAction<'app:hideKeyboard', void>
+export type MobileAppState = NoErrorTypedAction<'app:mobileAppState', {nextAppState: string}>
```
Comment (chrisnojima) [inline shared/constants/app.js]: we can more strongly type this

```
{nextAppState: 'active' | 'inactive' | 'background'}
```

Comment (mmaxim) [inline shared/constants/app.js]: Makes sense.

**Pull request merged**